### PR TITLE
Module 4 - Become a central banker - cash flow

### PR DESCRIPTION
**Pull request by DZariusz**
At this moment I wasn't able to test the solution, because I'm stuck with one issue:

How to execute flow command in node shel when one of arguments is a list:
`class Initiator(val linearIds: List<String>, ...` ?
I tried several combinations like `[.., ..]` {.., ..}`, `listOf(.., ..)` but nothing works so far.

### CODE REVIEW

File: contracts/src/main/kotlin/com/template/states/CashState.kt [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import com.template.contracts.CashContract
+import net.corda.core.contracts.BelongsToContract
+import net.corda.core.contracts.LinearState
+import net.corda.core.contracts.UniqueIdentifier
+import net.corda.core.identity.AbstractParty
+import net.corda.core.identity.Party
+import net.corda.core.schemas.MappedSchema
+import net.corda.core.schemas.PersistentState
+import net.corda.core.schemas.QueryableState
+
```
Comment (adelRestom) [inline contracts/src/main/kotlin/com/template/states/CashState.kt]: Cash state should have:
1. `issuer`: The bank that issues this currency. Cash issued by BankA is not the same as cash issued by BankB; you can't mix (i.e. merge and split) Cash states from different banks.
2. `holder`: The party that holds the amount. The bank issues the cash to a holder, one holder can transfer their cash to another holder.
3. `quantity`

File: contracts/src/main/kotlin/com/template/contracts/CashContract.kt [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+class CashContract : Contract {
+    companion object {
+        // Used to identify our contract when building a transaction.
+        val ID = CashContract::class.java.canonicalName
+    }
+
+    override fun verify(tx: LedgerTransaction) {
+        val command = tx.commands.requireSingleCommand<Commands>()
+
+        when (command.value) {
```
Comment (adelRestom) [inline contracts/src/main/kotlin/com/template/contracts/CashContract.kt]: Only the `issuer` needs to sign.

File: contracts/src/main/kotlin/com/template/contracts/CashContract.kt [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+            is Commands.Create -> requireThat {
```
Comment (adelRestom) [inline contracts/src/main/kotlin/com/template/contracts/CashContract.kt]: Only the `issuer` needs to sign.

File: contracts/src/main/kotlin/com/template/contracts/CashContract.kt [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+class CashContract : Contract {
+    companion object {
+        // Used to identify our contract when building a transaction.
+        val ID = CashContract::class.java.canonicalName
+    }
+
+    override fun verify(tx: LedgerTransaction) {
+        val command = tx.commands.requireSingleCommand<Commands>()
+
```
Comment (adelRestom) [inline contracts/src/main/kotlin/com/template/contracts/CashContract.kt]: Organize your verification rules in the following groups:
1. Shape of transaction (i.e. number of inputs/outputs).
2. Required signers.
3. Content of transaction (any business rule you can think of; like sum of inputs = sum of outputs).

File: contracts/src/main/kotlin/com/template/contracts/CashContract.kt [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+                val totalIn = tx.inputs.map { (it.state.data as CashState).value }.sum()
+                val totalOut = tx.outputs.map { (it.data as CashState).value }.sum()
+                "total input value must mach total output value" using (totalIn == totalOut)
+
+                val outputsParticipants = tx.outputsOfType<CashState>().flatMap { it.participants }
+                "All output participants must be signers." using (command.signers.containsAll(outputsParticipants.map { it.owningKey }))
+
+                if (tx.outputs.size == 2) {
+                    "receivers for both outputs must differ" using validateTransferWithChange(tx.outputs)
+                } else {
```
Comment (adelRestom) [inline contracts/src/main/kotlin/com/template/contracts/CashContract.kt]: One of the receivers can be the sender, imagine this scenario:
1. You have Cash state with amount 5
2. You only want to send amount 3 to PartyB
3. Your transaction will have 2 outputs:
    a. Cash state with amount 3 and holder PartyB
    b. The change: Cash state with amount 2 and holder is the sender

File: contracts/src/main/kotlin/com/template/contracts/CashContract.kt [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+                }
+
+            }
+            is Commands.Redeem -> requireThat {
+                "There must be at least one input." using (tx.inputs.size > 0)
+                "There must be one or no output" using (tx.outputs.size <= 1)
+                "Expect two signers." using (command.signers.size == 2)
+
+                "All inputs are from the same sender" using inputsHaveSameSender(tx.inputs)
+
```
Comment (adelRestom) [inline contracts/src/main/kotlin/com/template/contracts/CashContract.kt]: When you redeem, you are destroying the Cash state, there is no output.
This of it this way:
1. A bank issues a Cash state (token); this state is a promise from the bank to give you the equivalent of that state off-ledger when you redeem the state on-ledger.
2. Meaning when a bank issues you a Cash state with amount 5$, it's a promise that when you redeem that state on-ledger, the bank will give you in return real 5$ off-ledger, so when you redeem, the state is destroyed on-ledger; it's no longer tracked by Corda.

File: workflows/src/main/kotlin/com/template/flows/CashCreateFlow.kt [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+import net.corda.core.transactions.TransactionBuilder
+
+object CashCreateFlow {
+    @InitiatingFlow
+    @StartableByRPC
+    class Initiator(val value: Int) : FlowLogic<SignedTransaction>() {
+        @Suspendable
+        override fun call(): SignedTransaction {
+            val bankParty = BankProvider.call(serviceHub)
+
```
Comment (adelRestom) [inline workflows/src/main/kotlin/com/template/flows/CashCreateFlow.kt]: Why's that? Banks create the money, not people.

File: workflows/src/main/kotlin/com/template/flows/CashCreateFlow.kt [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+            val fullySignedTx = subFlow(CollectSignaturesFlow(
+                    signedTx,
+                    setOf(bankSession)
+            ))
+
+            return subFlow(FinalityFlow(fullySignedTx, setOf(bankSession)))
+        }
+    }
+
```
Comment (adelRestom) [inline workflows/src/main/kotlin/com/template/flows/CashCreateFlow.kt]: I don't think you need a responder or the signature of the holder, when money is issued to someone only the bank signs, when the money is transfered from one holder to the other; only the current holder signs, when money is redeemed both the bank and the holder must sign (bank must sign because the bank will give you the equivalent off-ledger, so they must agree on this transaction).

File: workflows/src/main/kotlin/com/template/flows/CashCreateFlow.kt [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    @InitiatedBy(Initiator::class)
+    class Acceptor(val cashCreatorSession: FlowSession) : FlowLogic<SignedTransaction>() {
+        @Suspendable
+        override fun call(): SignedTransaction {
+
+            val signTransactionFlow = object : SignTransactionFlow(cashCreatorSession) {
+                override fun checkTransaction(stx: SignedTransaction) = requireThat {
+                    val output = stx.tx.outputs[0].data as CashState
+
```
Comment (adelRestom) [inline workflows/src/main/kotlin/com/template/flows/CashCreateFlow.kt]: Not sure what you're checking here.

File: workflows/src/main/kotlin/com/template/flows/CashDestroyFlow.kt [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+import net.corda.core.identity.Party
+import net.corda.core.node.services.vault.QueryCriteria
+import net.corda.core.node.services.vault.QueryCriteria.LinearStateQueryCriteria
+import net.corda.core.transactions.SignedTransaction
+import net.corda.core.transactions.TransactionBuilder
+import java.util.*
+
+object CashDestroyFlow {
+    @InitiatingFlow
+    @StartableByRPC
```
Comment (adelRestom) [inline workflows/src/main/kotlin/com/template/flows/CashDestroyFlow.kt]: You know in the real world, the user won't know the linear Id's, instead they will say I want to redeem (distroy) 5$ issued by BankA; then you have to query the vault for all Cash states where issuer = BankA, holder = flow initiator, and quantity = 5, so you have to gather enough Cash states where the sum of of amount is >= 5, then your transaction might have an output (change) in the case the sum is greater than 5 (let's say 7), then you would have one output with amount 2 and holder as the initiator.  
So in short, change input parameters to `issuer` and `amount`, and inside the flow query the vault using the custom schema that you created to find the related Cash states.

File: workflows/src/main/kotlin/com/template/flows/CashDestroyFlow.kt [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+        override fun call(): SignedTransaction {
+            val bankParty = BankProvider.call(serviceHub)
+
+            if (ourIdentity.equals(bankParty)) {
+                throw IllegalArgumentException("Bank can't start Destroy flow")
+            }
+
+            val uuids: List<UUID> = linearIds.map { UUID.fromString(it) }
+            val queryCriteria: QueryCriteria = LinearStateQueryCriteria().withUuid(uuids)
+
```
Comment (adelRestom) [inline workflows/src/main/kotlin/com/template/flows/CashDestroyFlow.kt]: Btw, sometimes you might get more than 200 records in your results set and if you're not using pagination; that will throw an error: https://docs.corda.net/api-vault-query.html#pagination

File: workflows/src/main/kotlin/com/template/flows/CashDestroyFlow.kt [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+            var total = 0
+
+            // consume all provided inputs states and calculate total value
+            results.states.forEach {
+                val stateAndRef = serviceHub.toStateAndRef<ContractState>(it.ref)
+                val cashState = stateAndRef.state.data as CashState
+
+                txBuilder.addInputState(stateAndRef)
+
```
Comment (adelRestom) [inline workflows/src/main/kotlin/com/template/flows/CashDestroyFlow.kt]: Break the loop as soon as you reach the total that covers the required amount.

File: workflows/src/main/kotlin/com/template/flows/CashTransferFlow.kt [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+import net.corda.core.identity.Party
+import net.corda.core.node.services.vault.QueryCriteria
+import net.corda.core.node.services.vault.QueryCriteria.LinearStateQueryCriteria
+import net.corda.core.transactions.SignedTransaction
+import net.corda.core.transactions.TransactionBuilder
+import java.util.*
+
+object CashTransferFlow {
+    @InitiatingFlow
+    @StartableByRPC
```
Comment (adelRestom) [inline workflows/src/main/kotlin/com/template/flows/CashTransferFlow.kt]: Same idea, you should query the vault to find states that cover the required amount.

File: contracts/src/main/kotlin/com/template/contracts/CashContract.kt [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    override fun verify(tx: LedgerTransaction) {
+        val command = tx.commands.requireSingleCommand<Commands>()
+
+        when (command.value) {
+            is Commands.Create -> requireThat {
+                "Expect two signers." using (command.signers.size == 2)
+                "No inputs should be consumed" using (tx.inputs.isEmpty())
+                "Only one output state should be created." using (tx.outputs.size == 1)
```
Comment (xavierlepretre) [inline contracts/src/main/kotlin/com/template/contracts/CashContract.kt]: Same comment as the other CorDapp example PR: you are preventing mixing it with something else.

File: contracts/src/main/kotlin/com/template/contracts/CashContract.kt [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+
+                val out = tx.outputsOfType<CashState>().single()
+
+                "All parties must be signers." using (command.signers.containsAll(out.participants.map { it.owningKey }))
+                "The value must be non-negative." using (out.value > 0)
+            }
+            is Commands.Transfer -> requireThat {
+                "There must be at least one input." using (tx.inputs.size > 0)
+                "There must be one or two output(s) only" using (listOf(1, 2).contains(tx.outputs.size))
```
Comment (xavierlepretre) [inline contracts/src/main/kotlin/com/template/contracts/CashContract.kt]: 1 or 2? Too restrictive.

File: contracts/src/main/kotlin/com/template/contracts/CashContract.kt (hunk 3 above)
Comment (xavierlepretre) [reply contracts/src/main/kotlin/com/template/contracts/CashContract.kt]: You should comment below, that would be clearer. See ☝️ :

File: contracts/src/main/kotlin/com/template/states/CashState.kt (hunk 1 above)
Comment (xavierlepretre) [reply contracts/src/main/kotlin/com/template/states/CashState.kt]: If by `sender` you meant `issuer`, that was confusing. Your naming reminds me of a movement, a transaction, not of state.

File: contracts/src/main/kotlin/com/template/states/CashState.kt [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+@BelongsToContract(CashContract::class)
+data class CashState(val value: Int,
+                     val sender: Party,
+                     val receiver: Party,
+                     override val linearId: UniqueIdentifier = UniqueIdentifier()) :
+        LinearState, QueryableState {
```
Comment (xavierlepretre) [inline contracts/src/main/kotlin/com/template/states/CashState.kt]: What is the rationale for `LinearState`? You know that the onus is on the developer to ensure that the vault has a single un-consumed instance with a given id at any given time.

File: workflows/src/main/kotlin/com/template/flows/utils/BankProvider.kt [hunk 17]
```diff
@@ -0,0 +1,13 @@
...
+import net.corda.core.identity.CordaX500Name
+import net.corda.core.identity.Party
+import net.corda.core.node.ServiceHub
+
+object BankProvider {
+    fun call(serviceHub: ServiceHub) : Party {
+        val x500Name = CordaX500Name.parse("O=Bank,L=London,C=GB")
+        return serviceHub.networkMapCache.getPeerByLegalName(x500Name)
+                ?: throw IllegalArgumentException("Bank does not exist")
+    }
```
Comment (xavierlepretre) [inline workflows/src/main/kotlin/com/template/flows/utils/BankProvider.kt]: You could use Kotlin's receiver concept: https://stackoverflow.com/questions/45875491/what-is-a-receiver-in-kotlin

```kotlin
fun ServiceHub.getBankProvider() = networkMapCache
    .getPeerByLegalName(CordaX500Name.parse("O=Bank,L=London,C=GB"))
    ?: throw IllegalArgumentException("Bank does not exist")
```

To be called elsewhere as `serviceHub.getBankProvider()`.

Comment (xavierlepretre) [reply workflows/src/main/kotlin/com/template/flows/utils/BankProvider.kt]: And keep this well-known `"O=Bank,L=London,C=GB"` elsewhere more visible.

File: workflows/src/main/kotlin/com/template/flows/CashDestroyFlow.kt [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+            )
+
+            // build tx that will consume previous outputs
+            val txBuilder = TransactionBuilder(notary).addCommand(txCommand)
+
+            var total = 0
+
+            // consume all provided inputs states and calculate total value
+            results.states.forEach {
+                val stateAndRef = serviceHub.toStateAndRef<ContractState>(it.ref)
```
Comment (xavierlepretre) [inline workflows/src/main/kotlin/com/template/flows/CashDestroyFlow.kt]: `val stateAndRef = it`

File: workflows/src/main/kotlin/com/template/flows/CashDestroyFlow.kt [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+
+            val txCommand = Command(
+                    CashContract.Commands.Transfer(),
+                    listOf(ourIdentity.owningKey, bankParty.owningKey)
+            )
+
+            // build tx that will consume previous outputs
+            val txBuilder = TransactionBuilder(notary).addCommand(txCommand)
+
+            var total = 0
```
Comment (xavierlepretre) [inline workflows/src/main/kotlin/com/template/flows/CashDestroyFlow.kt]: `var`? Adel and I just shedded a tear... Here is how you make it a `val`. Take time to dig through this code:

```kotlin
val total = results.states.fold(0) { sum, stateAndRef ->
    txBuilder.addInputState(stateAndRef)
    sum + stateAndRef.state.data
            .also {
                require(it.receiver.equals(ourIdentity)) { "Out identity must own all inputs" }
            }
            .value
}
```

File: workflows/src/main/kotlin/com/template/flows/CashTransferFlow.kt [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+    class Initiator(val linearIds: List<String>, val value: Int, val receiver: Party) : FlowLogic<SignedTransaction>() {
+        @Suspendable
+        override fun call(): SignedTransaction {
+            val bankParty = BankProvider.call(serviceHub)
+
+            if (listOf(receiver, ourIdentity).contains(bankParty)) {
+                throw IllegalArgumentException("Bank can't be a party of Transfer flow")
```
Comment (xavierlepretre) [inline workflows/src/main/kotlin/com/template/flows/CashTransferFlow.kt]: It would be good to throw a `FlowException` instead.

File: workflows/src/main/kotlin/com/template/flows/CashTransferFlow.kt [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+                require(cashState.receiver.equals(ourIdentity)) { "Out identity must own all inputs" }
+            }
+
+            if (value < total) {
+                val changeState = CashState(total - value, ourIdentity, ourIdentity)
+                txBuilder.addOutputState(changeState, CashContract.ID)
+            }
+
+            val newState = CashState(value, ourIdentity, receiver)
+            txBuilder.addOutputState(newState, CashContract.ID)
```
Comment (xavierlepretre) [inline workflows/src/main/kotlin/com/template/flows/CashTransferFlow.kt]: Make the outputs more versatile, I want to pay 2 people and keep change for myself, for instance.
